circuit-benchmarks/src/pi_circuit.rs

```diff
@@ -0,0 +1,152 @@
+//! State circuit benchmarks
+
+#[cfg(test)]
+mod tests {
+    use ark_std::{end_timer, start_timer};
+    use eth_types::Word;
+    use halo2_proofs::arithmetic::Field;
+    use halo2_proofs::plonk::{create_proof, keygen_pk, keygen_vk, verify_proof};
+    use halo2_proofs::poly::kzg::commitment::{KZGCommitmentScheme, ParamsKZG, ParamsVerifierKZG};
+    use halo2_proofs::poly::kzg::multiopen::{ProverSHPLONK, VerifierSHPLONK};
+    use halo2_proofs::poly::kzg::strategy::SingleStrategy;
+    use halo2_proofs::{
+        halo2curves::bn256::{Bn256, Fr, G1Affine},
+        poly::commitment::ParamsProver,
+        transcript::{
+            Blake2bRead, Blake2bWrite, Challenge255, TranscriptReadBuffer, TranscriptWriterBuffer,
+        },
+    };
+    use rand::SeedableRng;
+    use rand_chacha::ChaCha20Rng;
+    use rand_xorshift::XorShiftRng;
+    use std::env::var;
+    use zkevm_circuits::pi_circuit::{
+        raw_public_inputs_col, PiCircuit, PublicData, BLOCK_LEN, EXTRA_LEN, TX_LEN,
+    };
+    use zkevm_circuits::test_util::rand_tx;
+    use zkevm_circuits::util::random_linear_combine_word as rlc;
+
+    #[cfg_attr(not(feature = "benches"), ignore)]
+    #[test]
+    fn bench_pi_circuit_prover() {
+        let degree: u32 = var("DEGREE")
+            .unwrap_or("15".to_string())
+            .parse()
+            .expect("Cannot parse DEGREE env var as u32");
+
+        const MAX_TXS: usize = 10;
+
+        const MAX_CALLDATA: usize = 128;
+
+        // Initialize the polynomial commitment parameters
+        let mut rng = XorShiftRng::from_seed([
+            0x59, 0x62, 0xbe, 0x5d, 0x76, 0x3d, 0x31, 0x8d, 0x17, 0xdb, 0x37, 0x32, 0x54, 0x06,
+            0xbc, 0xe5,
+        ]);
+
+        // Create the circuit
+        let randomness = <Fr as Field>::random(&mut rng);
+        let public_data = fillup_publicdata::<MAX_TXS, MAX_CALLDATA>();
+
+        let rlc_rpi_col =
+            raw_public_inputs_col::<Fr, MAX_TXS, MAX_CALLDATA>(&public_data, randomness);
+        assert_eq!(
+            rlc_rpi_col.len(),
+            BLOCK_LEN + 1 + EXTRA_LEN + 3 * (TX_LEN * MAX_TXS + 1 + MAX_CALLDATA)
+        );
+
+        // Computation of raw_pulic_inputs
+        let rlc_rpi = rlc_rpi_col
+            .iter()
+            .rev()
+            .fold(Fr::zero(), |acc, val| acc * randomness + val);
+
+        let public_inputs = vec![
+            randomness,
+            rlc_rpi,
+            Fr::from(public_data.extra.chain_id.as_u64()),
+            rlc(
+                public_data.extra.eth_block.state_root.to_fixed_bytes(),
+                randomness,
+            ),
+            rlc(public_data.prev_state_root.to_fixed_bytes(), randomness),
+        ];
+
+        let circuit = PiCircuit::<Fr, MAX_TXS, MAX_CALLDATA> {
+            randomness: randomness,
+            rand_rpi: randomness.clone(),
+            public_data: public_data,
+        };
```

Comment:
Is it possible to move this witness generation code to the public input circuit itself like here: https://github.com/privacy-scaling-explorations/zkevm-circuits/blob/77c5b62ee6c29ec4a2ae811a18fc81a6a64a0c2e/circuit-benchmarks/src/bit_keccak.rs#L36? There's still a lot of circuit details exposed here so if we can move them to the implementation file itself that seems to be less error prone.

Reply:
Yes, should be! Code updated!